bitcoin/src/bip32.rs

```diff
@@ -111,40 +111,77 @@ pub struct Xpub {
 #[cfg(feature = "serde")]
 internals::serde_string_impl!(Xpub, "a BIP-32 extended public key");
 
-/// A child number for a derived key
+/// A BIP32 types of key derivation
 #[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
 pub enum ChildNumber {
-    /// Non-hardened key
-    Normal {
-        /// Key index, within [0, 2^31 - 1]
-        index: u32,
-    },
-    /// Hardened key
-    Hardened {
-        /// Key index, within [0, 2^31 - 1]
-        index: u32,
-    },
+    /// Enum wrapper for a normal child number
+    Normal(NormalChildNumber),
+    /// Enum wrapper for a hardened child number
+    Hardened(HardenedChildNumber),
+}
+
+/// A child number for a derived key
+/// Indexes 2147483648 to 4294967295 (the second half of all possible children)
+#[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+pub struct HardenedChildNumber(u32);
+
+impl HardenedChildNumber {
+    /// `HardenedChildNumber` with index 0.
+    pub const ZERO: Self = HardenedChildNumber(0);
+    /// `HardenedChildNumber` with index 1.
+    pub const ONE: Self = HardenedChildNumber(1);
+}
+
+/// NormalChildNumber is a ChildNumber that is not hardened
+/// Indexes 0 to 2147483647 (the first half of all possible children).
+#[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+pub struct NormalChildNumber(u32);
+impl NormalChildNumber {
+    /// `NormalChildNumber` with index 0.
+    pub const ZERO: Self = NormalChildNumber(0);
+    /// `NormalChildNumber` with index 1.
+    pub const ONE: Self = NormalChildNumber(1);
+}
+impl From<u32> for NormalChildNumber {
+    fn from(num: u32) -> Self { NormalChildNumber(num) }
+}
+impl From<u32> for HardenedChildNumber {
+    fn from(num: u32) -> Self { HardenedChildNumber(num) }
```

Comment:
Forgot about the constraints, sorry. Ill implement by the morning.

Reply:
You can just delete those two.